agent/hcp/client/metrics_client_test.go

```diff
@@ -0,0 +1,162 @@
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"golang.org/x/oauth2"
+	"google.golang.org/protobuf/proto"
+
+	colpb "go.opentelemetry.io/proto/otlp/collector/metrics/v1"
+	metricpb "go.opentelemetry.io/proto/otlp/metrics/v1"
+
+	"github.com/hashicorp/consul/version"
+	"github.com/hashicorp/go-hclog"
+	hcpcfg "github.com/hashicorp/hcp-sdk-go/config"
+	"github.com/stretchr/testify/require"
+)
+
+type mockHCPCfg struct{}
+
+func (m *mockHCPCfg) Token() (*oauth2.Token, error) {
+	return &oauth2.Token{
+		AccessToken: "test-token",
+	}, nil
+}
+
+func (m *mockHCPCfg) APITLSConfig() *tls.Config { return nil }
+
+func (m *mockHCPCfg) SCADAAddress() string { return "" }
+
+func (m *mockHCPCfg) SCADATLSConfig() *tls.Config { return &tls.Config{} }
+
+func (m *mockHCPCfg) APIAddress() string { return "" }
+
+func (m *mockHCPCfg) PortalURL() *url.URL { return &url.URL{} }
+
+type mockCloudCfg struct{}
+
+func (m mockCloudCfg) HCPConfig(opts ...hcpcfg.HCPConfigOption) (hcpcfg.HCPConfig, error) {
+	return &mockHCPCfg{}, nil
+}
+
+type mockErrCloudCfg struct{}
+
+func (m mockErrCloudCfg) HCPConfig(opts ...hcpcfg.HCPConfigOption) (hcpcfg.HCPConfig, error) {
+	return nil, errors.New("test bad HCP config")
+}
+
+func TestNewMetricsClient(t *testing.T) {
+	for name, test := range map[string]struct {
+		wantErr string
+		cfg     *TelemetryClientCfg
+	}{
+		"success": {
+			cfg: &TelemetryClientCfg{
+				Logger:   hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+				CloudCfg: &mockCloudCfg{},
+			},
+		},
+		"failsWithoutCloudCfg": {
+			wantErr: "failed to init telemetry client",
+			cfg: &TelemetryClientCfg{
+				Logger:   hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+				CloudCfg: nil,
+			},
+		},
+		"failsWithoutLogger": {
+			wantErr: "failed to init telemetry client",
+			cfg: &TelemetryClientCfg{
+				Logger:   nil,
+				CloudCfg: &mockErrCloudCfg{},
+			},
+		},
+		"failsHCPConfig": {
+			wantErr: "failed to init telemetry client",
+			cfg: &TelemetryClientCfg{
+				Logger:   hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+				CloudCfg: &mockErrCloudCfg{},
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			client, err := NewMetricsClient(test.cfg)
+			if test.wantErr != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), test.wantErr)
+				return
+			}
+
+			require.Nil(t, err)
+			require.NotNil(t, client)
+		})
+	}
+}
+
+func TestExportMetrics(t *testing.T) {
+	for name, test := range map[string]struct {
+		wantErr string
+		status  int
+	}{
+		"success": {
+			status: http.StatusOK,
+		},
+		"failsWithNonRetryableError": {
+			status:  http.StatusBadRequest,
```

Comment:
Will partial success always return badrequest?

Reply:
Not necessarily actually, I just added it to test out that part of my code which handles the partial success